### PR TITLE
Fork tests for liquidation

### PR DESCRIPTION
**Pull request by GitGuru7**
## Description

It includes fork tests of liquidation through VToken

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: tests/hardhat/Fork/liquidation.ts [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+      });
+
+      it("liquidate user", async () => {
+        await comptroller.setMinLiquidatableCollateral(0);
+        await priceOracle.setDirectPrice(usdd.address, convertToUnit("100", 15));
+        const borrowBalance = await vUSDT.borrowBalanceStored(acc2);
+        const closeFactor = await comptroller.closeFactorMantissa();
+        const maxClose = (borrowBalance * closeFactor) / 1e18;
+        const result = vUSDT.connect(acc1Signer).liquidateBorrow(acc2, maxClose.toString(), vUSDD.address);
+        await expect(result).to.emit(vUSDT, "LiquidateBorrow");
```
Comment (chechu) [inline tests/hardhat/Fork/liquidation.ts:191]: I would check more values:
* new borrow balance of acc2 // it should be lower
* new vUSDT balance of acc2 // it should be lower
* new vUSDT balance of acc1Signer // it should be higher
* new reserves in the vUSDT market // it should be higher

I would expect exact numbers, not only lower or greater checks

File: tests/hardhat/Fork/liquidation.ts [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        ).to.be.revertedWithCustomError(comptroller, "TooMuchRepay");
+      });
+
+      it("liquidate user", async () => {
+        await comptroller.setMinLiquidatableCollateral(0);
+        await priceOracle.setDirectPrice(usdd.address, convertToUnit("100", 15));
+        const borrowBalance = await vUSDT.borrowBalanceStored(acc2);
+        const closeFactor = await comptroller.closeFactorMantissa();
+        const maxClose = (borrowBalance * closeFactor) / 1e18;
+        const result = vUSDT.connect(acc1Signer).liquidateBorrow(acc2, maxClose.toString(), vUSDD.address);
```
Comment (chechu) [inline tests/hardhat/Fork/liquidation.ts:190]: Before liquidating I would check Comptroller.getAccountLiquidity, to be sure the account is underwater according to that function

File: tests/hardhat/Fork/liquidation.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+      await configureTimelock();
+
+      acc1Signer = await initMainnetUser(acc1, ethers.utils.parseUnits("2"));
+      acc2Signer = await initMainnetUser(acc2, ethers.utils.parseUnits("2"));
+
+      usdt = FaucetToken__factory.connect("0xA11c8D9DC9b66E209Ef60F0C8D969D3CD988782c", impersonatedTimelock);
+      usdd = MockToken__factory.connect("0x2E2466e22FcbE0732Be385ee2FBb9C59a1098382", impersonatedTimelock);
+      vUSDT = await configureVToken("0x296da137120562c79b26808c1aa142a59ebf31f4");
+      vUSDD = await configureVToken("0xeD7401294EBF0A1b0721562a69031565F4a4Bacd");
+      comptroller = Comptroller__factory.connect("0x605AA769d14F6Af2E405295FEC2A4d8Baa623d80", impersonatedTimelock);
```
[the remark this answers is not in the document]
Comment (chechu) [reply tests/hardhat/Fork/liquidation.ts:99]: @0xlucian, do you think we could maintain these addresses in a deployment JSON file? Maybe we would have to maintain them manually, right? Let's try to talk about this offline

File: tests/hardhat/Fork/liquidation.ts [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+        );
+      });
+
+      it("Should success on liquidation when repay amount is equal to borrowing", async function () {
+        await usdd.connect(acc2Signer).faucet(10900000000);
+        await usdd.connect(acc2Signer).approve(vUSDD.address, 10900000000);
+        await vUSDD.connect(acc2Signer).mint(10900000000);
+
+        await priceOracle.setDirectPrice(usdd.address, convertToUnit("1", 14)); // 100000000000000
+        await priceOracle.setDirectPrice(usdt.address, convertToUnit("1", 2)); // 100000000000000
```
Comment (chechu) [inline tests/hardhat/Fork/liquidation.ts:289]: Is the comment correct?

File: tests/hardhat/Fork/liquidation.ts [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+        await comptroller.setMinLiquidatableCollateral(0);
+        await priceOracle.setDirectPrice(usdd.address, convertToUnit("100", 15));
+        const borrowBalance = await vUSDT.borrowBalanceStored(acc2);
+
+        const [err, liquidity, shortfall] = await comptroller.getAccountLiquidity(acc2);
+        expect(err).equals(0);
+        expect(liquidity).equals(0);
+        expect(shortfall).greaterThan(0);
+
+        const totalRerservesUsddPrev = await vUSDD.totalReserves();
```
Comment (chechu) [inline tests/hardhat/Fork/liquidation.ts:201]: Typo. There are more instances

File: tests/hardhat/Fork/liquidation.ts [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        Calculations
+        snapshot.totalCollateral 1e9  // (bnxPrice * mint amount) / mantissa
+        snapshot.borrows 1e11    //  (BTCBPrice * BTCBBorrowAmount) / mantissa
+        percantage 0.1   (collateral/borrow) * mantissa
+        repaymentAmount 1000       percentage*borrowBalance 
+        borrowBalance 10000
+        */
+        const collateralAmount = 1e9;
+        const borrowAmount = 1e11;
+        const percantageOfRepay = (collateralAmount / borrowAmount) * 1e18;
```
Comment (chechu) [inline tests/hardhat/Fork/liquidation.ts:375]: Typo. `percantageOfRepay` vs `percentageOfRepay`

File: tests/hardhat/Fork/liquidation.ts [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+        /*
+        Calculations
+        snapshot.totalCollateral 1e9  // (bnxPrice * mint amount) / mantissa
+        snapshot.borrows 1e11    //  (BTCBPrice * BTCBBorrowAmount) / mantissa
+        percantage 0.1   (collateral/borrow) * mantissa
+        repaymentAmount 1000       percentage*borrowBalance 
+        borrowBalance 10000
+        */
+        const collateralAmount = 1e9;
+        const borrowAmount = 1e11;
```
Comment (chechu) [inline tests/hardhat/Fork/liquidation.ts:374]: Magic numbers?

**Review by chechu — COMMENTED** [review]
Tbh, for me, it would be easier to follow the rationale with a fixed price of $1 for usdt and usdd, and using different amounts to borrow and mint

File: contracts/test/Mocks/MockPriceOracle.sol [hunk 8]
```diff
@@ -2,6 +2,7 @@
 pragma solidity ^0.8.10;
 
 import { ResilientOracleInterface } from "@venusprotocol/oracle/contracts/interfaces/OracleInterface.sol";
+import { ChainlinkOracle } from "@venusprotocol/oracle/contracts/oracles/ChainlinkOracle.sol";
```
[the remark this answers is not in the document]
Comment (GitGuru7) [reply contracts/test/Mocks/MockPriceOracle.sol:5]: To create typechain for the chainlink oracle used in tests

File: tests/hardhat/Fork/liquidation.ts [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+  await tx.wait();
+
+  tx = await accessControlManager
+    .connect(impersonatedTimelock)
+    .giveCallPermission(comptroller.address, "setMinLiquidatableCollateral(uint256)", ADMIN);
+  await tx.wait();
+
+  tx = await accessControlManager
+    .connect(impersonatedTimelock)
+    .giveCallPermission(comptroller.address, "setCollateralFactor(address,uint256,uint256)", ADMIN);
```
[the remark this answers is not in the document]
Comment (GitGuru7) [reply tests/hardhat/Fork/liquidation.ts]: Fixed [here](https://github.com/VenusProtocol/isolated-pools/pull/227/commits/3fd29dac299b786124541ac84b70cda1266cdd6a)

File: tests/hardhat/Fork/liquidation.ts [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+          vUSDT.connect(acc1Signer).liquidateBorrow(acc2, usdtBorrowAmount, vUSDT.address),
+        ).to.be.revertedWithCustomError(comptroller, "InsufficientShortfall");
+      });
+
+      it("Should revert when liquidation is called through vToken and trying to seize more tokens", async function () {
+        await comptroller.setMinLiquidatableCollateral(0);
+        // Mint and Increase collateral of the user
+        await priceOracle.setDirectPrice(usdd.address, convertToUnit("1", 5));
+        // Liquidation
+        await expect(vUSDT.connect(acc1Signer).liquidateBorrow(acc2, 201, vUSDD.address)).to.be.revertedWith(
```
[the remark this answers is not in the document]
Comment (GitGuru7) [reply tests/hardhat/Fork/liquidation.ts]: Fixed [here](https://github.com/VenusProtocol/isolated-pools/pull/227/commits/3fd29dac299b786124541ac84b70cda1266cdd6a)

File: tests/hardhat/Fork/liquidation.ts [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+        const result = vUSDT.connect(acc1Signer).liquidateBorrow(acc2, maxClose.toString(), vUSDD.address);
+
+        await expect(result).to.emit(vUSDT, "LiquidateBorrow");
+        const vUSDDBalAcc2New = await vUSDD.balanceOf(acc2);
+        const vUSDDBalAcc1New = await vUSDD.balanceOf(acc1);
+        const totalRerservesUsddNew = await vUSDD.totalReserves();
+        const exchangeRateCollateralNew = await vUSDD.exchangeRateStored();
+
+        const liquidatorSeizeTokens = Math.floor((seizeTokens * 95) / 100);
+        const protocolSeizeTokens = Math.floor((seizeTokens * 5) / 100);
```
[the remark this answers is not in the document]
Comment (GitGuru7) [reply tests/hardhat/Fork/liquidation.ts]: Fixed [here](https://github.com/VenusProtocol/isolated-pools/pull/227/commits/3fd29dac299b786124541ac84b70cda1266cdd6a)

File: tests/hardhat/Fork/liquidation.ts (hunk 6 above)
Comment (GitGuru7) [reply tests/hardhat/Fork/liquidation.ts]: Fixed [here](https://github.com/VenusProtocol/isolated-pools/pull/227/commits/3fd29dac299b786124541ac84b70cda1266cdd6a)

File: tests/hardhat/Fork/liquidation.ts (hunk 7 above)
Comment (GitGuru7) [reply tests/hardhat/Fork/liquidation.ts]: Fixed [here](https://github.com/VenusProtocol/isolated-pools/pull/227/commits/3fd29dac299b786124541ac84b70cda1266cdd6a)

**Review by defcon022 — APPROVED** [review]

**Review by chechu — APPROVED** [review]

**Pull request merged**